server/src/test/java/org/elasticsearch/snapshots/SnapshotsServiceTests.java

```diff
@@ -246,9 +326,133 @@ public void testConcurrentSnapshotCreateAndDelete() {
+                                clusterStateResponse -> {
+                                    final ShardRouting shardToRelocate =
+                                        clusterStateResponse.getState().routingTable().allShards(index).get(0);
+                                    final TestClusterNode currentPrimaryNode =
+                                        testClusterNodes.nodeById(shardToRelocate.currentNodeId());
+                                    final TestClusterNode otherNode =
+                                        testClusterNodes.randomDataNodeSafe(currentPrimaryNode.node.getName());
+                                    final Runnable maybeForceAllocate = new Runnable() {
+                                        @Override
+                                        public void run() {
+                                            masterAdminClient.cluster().state(new ClusterStateRequest(), assertNoFailureListener(
+                                                resp -> {
+                                                    final ShardRouting shardRouting = resp.getState().routingTable()
+                                                        .shardRoutingTable(shardToRelocate.shardId()).primaryShard();
+                                                    if (shardRouting.unassigned()
+                                                        && shardRouting.unassignedInfo().getReason() == UnassignedInfo.Reason.NODE_LEFT) {
+                                                        if (masterNodeCount > 1) {
+                                                            scheduleNow(() -> testClusterNodes.stopNode(masterNode));
+                                                        }
+                                                        testClusterNodes.randomDataNodeSafe().client.admin().cluster()
+                                                            .prepareCreateSnapshot(repoName, snapshotName)
+                                                            .execute(ActionListener.wrap(() -> {
+                                                                testClusterNodes.randomDataNodeSafe().client.admin().cluster()
+                                                                    .deleteSnapshot(
+                                                                        new DeleteSnapshotRequest(repoName, snapshotName), noopListener());
+                                                                createdSnapshot.set(true);
+                                                            }));
+                                                        scheduleNow(
+                                                            () -> testClusterNodes.randomMasterNodeSafe().client.admin().cluster().reroute(
+                                                                new ClusterRerouteRequest().add(
+                                                                    new AllocateEmptyPrimaryAllocationCommand(
```

Comment:
Poor man's primary relocation simulation until we have non-blocking recoveries.

Reply:
what a hack :)